lib/services/local/embeddings.go

```diff
@@ -0,0 +1,178 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package local
+
+import (
+	"context"
+	"crypto/sha256"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport/api/utils/retryutils"
+	"github.com/gravitational/teleport/lib/ai"
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/teleport/lib/services"
+)
+
+// EmbeddingsService implements the services.Embeddings interface.
+type EmbeddingsService struct {
+	log    *logrus.Entry
+	jitter retryutils.Jitter
+	backend.Backend
+	client ai.Embedder
+	clock  clockwork.Clock
+}
+
+const (
+	embeddingsPrefix = "embeddings"
+	embeddingExpiry  = 30 * 24 * time.Hour // 30 days
+)
+
+// GetEmbedding looks up a single embedding by its name in the backend.
+func (e EmbeddingsService) GetEmbedding(ctx context.Context, kind, resourceID string) (*ai.Embedding, error) {
+	result, err := e.Get(ctx, backend.Key(embeddingsPrefix, kind, resourceID))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return services.UnmarshalEmbedding(result.Value)
+}
+
+// GetEmbeddings returns all embeddings for a given kind.
+func (e EmbeddingsService) GetEmbeddings(ctx context.Context, kind string) ([]*ai.Embedding, error) {
+	startKey := backend.Key(embeddingsPrefix, kind)
+	result, err := e.GetRange(ctx, startKey, backend.RangeEnd(startKey), backend.NoLimit)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	embeddings := make([]*ai.Embedding, len(result.Items))
+	var embedding *ai.Embedding
+	if len(result.Items) == 0 {
+		return nil, nil
+	}
+	for i, item := range result.Items {
+		embedding, err = services.UnmarshalEmbedding(item.Value)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		embeddings[i] = embedding
+	}
+	return embeddings, nil
+}
+
+// UpsertEmbedding creates or update a single ai.Embedding in the backend.
+func (e EmbeddingsService) UpsertEmbedding(ctx context.Context, embedding *ai.Embedding) (*ai.Embedding, error) {
+	value, err := services.MarshalEmbedding(embedding)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	_, err = e.Put(ctx, backend.Item{
+		Key:     embeddingItemKey(embedding),
+		Value:   value,
+		Expires: e.clock.Now().Add(embeddingExpiry),
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return embedding, nil
+}
+
+// Embed takes a resource textual representation, checks if the resource
+// already has an up-to-date embedding stored in the backend, and computes
+// a new embedding otherwise. The newly computed embedding is stored in
+// the backend.
+func (e EmbeddingsService) Embed(ctx context.Context, kind string, resources map[string][]byte) ([]*ai.Embedding, error) {
```

Comment:
`lib/services/local` doesn't feel like the correct place for this - so far the services have only had to deal with CPU-bound marshaling and unmarshaling and interacting with the backend, not call to some external service.

Reply:
I moved the logic out of the Embeddings service. It only performs backend queries and marshalling operations.